File: ruby/aoc.rb
#!/usr/bin/env ruby
# frozen_string_literal: true

require 'uri'
require 'net/http'
require 'fileutils'
require 'pastel'

pastel = Pastel.new

def get_input(year, day)
  path = "data/#{year}/#{day}.txt"
  FileUtils.mkdir_p "data/#{year}"
  return File.read(path) if File.exist? path

  cookie = File.read('.cookie').strip

  uri = URI("https://adventofcode.com/#{year}/day/#{day}/input")
  request = Net::HTTP::Get.new(uri)
  request['Cookie'] = "session=#{cookie}"
  request['User-Agent'] = 'https://github.com/d-nery/aoc <danielnso97@gmail.com>'
  input = Net::HTTP.start(uri.hostname, uri.port, use_ssl: uri.scheme == 'https') do |http|
    http.request(request)
  end.body.strip
  File.write(path, input)
  input
end

problem = ARGV[0]
day = ARGV[1] || Time.now.day
year = ARGV[2] || Time.now.year

require "./#{year}/day#{day}"

puts '== AoC Problem Runner =='
puts "Day #{pastel.yellow(day)}"
puts "Year #{pastel.yellow(year)}"
puts "Problem #{pastel.yellow(problem || 'test')}"
puts '------------------------'
puts

puts pastel.dim('Getting input...')
input = get_input(year, day)
puts pastel.dim('Done!')
puts '------------------------'
puts

puts 'Running...'

start = Process.clock_gettime(Process::CLOCK_MONOTONIC)
result = if problem == '1'
           part_one input
         elsif problem == '2'
           part_two input
         else
           test
           exit
         end
finish = Process.clock_gettime(Process::CLOCK_MONOTONIC)

puts "Done in #{((finish - start) * 1000).round(1)}ms! Result #{pastel.green(result)}"

File: bun/aoc.ts
#! /usr/bin/env bun

const problem = Bun.argv[2] ?? "test";
const day = Bun.argv[3] ?? new Date().getDate();
const year = Bun.argv[4] ?? new Date().getFullYear();

const fetch_input = async () => {
  const dataPath = `data/${year}/${day}.txt`;
  const file = Bun.file(dataPath);
  if (await file.exists()) {
    return (await file.text()).trim();
  }

  console.log("Fetching input...");
  const cookie = await Bun.file(".cookie").text();
  const request = new Request(
    `https://adventofcode.com/${year}/day/${day}/input`,
    {
      headers: {
        Cookie: `session=${cookie.trim()}`,
        "User-Agent": "https://github.com/d-nery/aoc <danielnso97@gmail.com>",
      },
    }
  );

  const raw = await fetch(request);
  const input = await raw.text();

  file.write(input).catch(console.error);

  return input.trim();
};

console.log("== AoC Problem Runner ==");
console.log(`Day ${day}`);
console.log(`Year ${year}`);
console.log(`Problem ${problem}`);
console.log("------------------------");
console.log();

const { partOne, partTwo, test } = await import(`./${year}/${day}.ts`);

const input = await fetch_input();

console.log("Running...");
console.log("------------------------");
console.log();
const start = performance.now();
let answer: any = 0;
if (problem === "1") {
  answer = partOne(input);
} else if (problem === "2") {
  answer = partTwo(input);
} else {
  test();
  process.exit(0);
}

const end = performance.now();

console.log(`Finished in ${(end - start).toFixed(3)} ms`);
console.log(`Answer: ${answer}`);
